Enhancement/issue 5 docker workflow

thescientist13 opened the pull request:
resolves #5 and some of #2 

1. Added a Dockerfile that is based on 
    - Debian Linux
    - Anaconda w/ Python3
    - Jupyter
1. Added general stack documentation to the README
1. Added Docker specific instructions and example workflow
1. Using `docker-compose` to help improve contributor /  CLI ergonomics
1. Version controlled a configuration file for the Jupyter Notebook server
1. Created a _workspace_ directory that can be used as a two-way mount point between the host and container
1. Make _workspace_ the working directory, where "starter" Jupyter Notebooks for contributors can be saved

<img width="1181" alt="screen shot 2018-01-01 at 11 29 22 am" src="https://user-images.githubusercontent.com/895923/34469345-e78fc4c4-eeea-11e7-9b5c-99ee7cbb334e.png">

```shell
root@2495f84d456d:/workspace# ls -l
total 8
-rw-r--r-- 1 root root 231 Jan  1 16:26 readme.txt
-rw-r--r-- 1 root root  12 Jan  1 16:29 test.txt
root@2495f84d456d:/workspace# exit
exit
Owens-MacBook-Pro:data-night owenbuckley$ git status
On branch enhancement/issue-5-docker-workflow
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	workspace/test.txt

nothing added to commit but untracked files present (use "git add" to track)
```

TODO
1. ~~Fix errors encountered by testing Docker steps~~
1. ~~Use `docker-compose` and _docker-compose.yml_ to minimize needed CLI arguments and to add general structure to Docker management / usage.  This will also scale nicely if we want to additional containers like for a database.  [see this for an example](https://github.com/thescientist13/game-of-maps/blob/enhancement/issue-2-guide-part-one/docker-compose.yml)~~
1. ~~Create a host specific mount point to be used with the Jupyter server that will persist user data after container exit~~
1.  Try and fix the `0.0.0.0` / `127.0.0.0` issue between OSX / Windows when loading the Jupyter Notebook server from the host

Commit:
WIP

Commit:
docker container with all tools and documented in README

Commit:
tweaks

Commit:
tweaks

Commit:
typo

paulopperman commented on Dockerfile:
```diff
@@ -0,0 +1 @@
+FROM continuumio/anaconda3:5.0.1
```
This should either clone the repo from github, or build from the local directory so we have things prepopulated.

thescientist13 replied:
not sure what you mean here?

paulopperman commented:
one thing we may have to do is get data into the container from the web or external source.  what would the workflow for that look like?

paulopperman reviewed (changes requested):
I wasn't able to get to the jupyter server running on the container from my local browser (kept getting connection reset error).  I think you need to do some additional configuration of the jupyter notebook server to allow external connections.

http://jupyter-notebook.readthedocs.io/en/stable/public_server.html

thescientist13 commented:
Good question!

Via a web source, just use something like `wget` when in the container.  

For general purpose file access though, likely it's just a simple matter of mounting some sort of "data" directory in the repo to the container on start as a volume, which can then be accessed and  via the container but also persisted .
https://docs.docker.com/engine/admin/volumes/volumes/

I will get an example of that in this PR.

thescientist13 commented:
@paulopperman 
> I wasn't able to get to the jupyter server running on the container from my local browser (kept getting connection reset error). I think you need to do some additional configuration of the jupyter notebook server to allow external connections.

I hadn't included the hub feature as part of this effort, just the notebook as I thought we were just going to use [jupyterhub-deploy-docker](https://github.com/jupyterhub/jupyterhub-deploy-docker) on a droplet?

Can you confirm if running notebook worked for you though?

paulopperman commented:
@thescientist13 the issue is from following the README instructions to launch jupyter notebook.  the container runs fine, but couldn't get a browser to connect to the server.  i'll show my workaround on the file.

paulopperman commented on README.md:
```diff
@@ -43,12 +43,41 @@ We've set up a few different possible options for working with the data.  You ca
+1. [Install Docker](https://www.docker.com/community-edition).
+1. In the command shell, navigate to the root of this repository
+1. Run `docker build -t data-night .`  (First time running this may a take a little time)
+1. Now you can connect to your container via a shell `docker run -p 8888:8888 -it sandbox`
```
fix the container name, and map a different port to avoid conflicting with a local jupyter server:
`docker run -p 8889:8888 -it data-night`

paulopperman commented on README.md:
````diff
@@ -43,12 +43,41 @@ We've set up a few different possible options for working with the data.  You ca
+ 
+For example, to start a jupyter notebook locally, run:
+```shell
+:/# jupyter notebook --ip=127.0.0.1 --allow-root
````
@thescientist13 i had to do thre things to get this to work:
1. 
```# jupyter notebook --generate-config --allow-root```

2.
edit `~/.jupyter/jupyter_notebook_config.py` to change `c.NotebookApp.ip = '*'`

3.
from home, run `jupyter notebook --allow-root`

after that, i can get to the notebook server from my local machine

thescientist13 replied:
hmm, curious, what was the error / console output you saw when running
```shell
jupyter notebook --ip=127.0.0.1 --allow-root
```
?

paulopperman replied:
No console error. Browser error (connection refused)

thescientist13 replied:
i see.  I just realized i had the jupyter server we setup from the other daty still running and might have been connecting to that inadvertently 🤦‍♂️ .  

Trying my steps over, which will likely me see me reproducing what you saw.

Commit:
updated Docker instructions

thescientist13 commented:
@paulopperman 
I updated the Docker steps (mind giving it another shot?) and updated the issue description with TODOs.  Shouldn't be much more effort to get this all wired up nice-nice.

Commit:
updated Docker instructions

davenamin reviewed (commented):
this looks good! was able to follow all the steps with Docker on Windows, but one hiccup and one suggestion:

the line "And visit the generated URL after the server starts" isn't quite right, because trying to navigate to 0.0.0.0 on the host fails. It does work if I manually change it to 127.0.0.1:8889...blahblahblah, but I have a suggestion...

why not put the "jupyter notebook --ip=...." command into the Dockerfile explicitly? given, that wouldn't work so well with the default random token needed to log into the notebook, but setting a default password might make things a lot easier. that would require creating a configuration file (like what @paulopperman had suggested when running into the IP address issue) but would simplify things for someone who isn't familiar with Docker or CLI usage. the configuration file can be checked into the repo and ADDed into the Docker container during the build steps.

thescientist13 commented:
@davenamin 
Thanks for the reviewing and providing feedback!  

> "And visit the generated URL after the server starts" isn't quite right, because trying to navigate to 0.0.0.0 on the host fails. It does work if I manually change it to 127.0.0.1:8889...blahblahblah, but I have a suggestion...

Ah, good find.  Maybe `0.0.0.0` presents an issue on Windows, will see if `127.0.0.1` works even if binding to `0.0.0.0` when starting Jupyter Notebook.

> why not put the "jupyter notebook --ip=...." command into the Dockerfile explicitly?

Agreed!  That's my next task with this PR.  (You can see what I plan to do specifically in item #2 from TODO section of this PR description)

davenamin commented:
> Ah, good find. Maybe 0.0.0.0 presents an issue on Windows, will see if 127.0.0.1 works even if binding to 0.0.0.0 when starting Jupyter Notebook.

Sorry, wasn't very specific with my language: binding to 0.0.0.0 definitely works on Windows (and I'd guess OSX and Linux) but when the `jupyter notebook --ip=0.0.0.0 --port 8889 --allow-root` command finishes and prints the URL to use, that URL isn't actually valid because it contains the wildcard IP. The README says:

>And visit the generated URL after the server starts, e.g. http://0.0.0.0:8889/?token=39f8a00b806723e69e34ce0c471165100453f3f8f80a2126

but after pasting the generated URL into a browser, it won't work until the beginning IP is changed to 127.0.0.1 (or `localhost`).

paulopperman commented:
I'm offline, but earlier, for the IP, I could just go to localhost:8888 on
my machine to get to the notebook. I didn't run the server with the IP
specified though.

On Dec 31, 2017 7:07 PM, "davenamin" <notifications@github.com> wrote:

> Ah, good find. Maybe 0.0.0.0 presents an issue on Windows, will see if
> 127.0.0.1 works even if binding to 0.0.0.0 when starting Jupyter Notebook.
>
> Sorry, wasn't very specific with my language: binding to 0.0.0.0
> definitely works on Windows (and I'd guess OSX and Linux) but when the jupyter
> notebook --ip=0.0.0.0 --port 8889 --allow-root command finishes and
> prints the URL to use, that URL isn't actually valid because it contains
> the wildcard IP. The README says:
>
> And visit the generated URL after the server starts, e.g.
> http://0.0.0.0:8889/?token=39f8a00b806723e69e34ce0c471165
> 100453f3f8f80a2126
>
> but after pasting the generated URL into a browser, it won't work until
> the beginning IP is changed to 127.0.0.1 (or localhost).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NewportDataProject/data-night/pull/8#issuecomment-354629280>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGAeGkC16Ttu6M5bC2xckfPniaIfm7Uuks5tGCGlgaJpZM4RP5zx>
> .
>

Commit:
updated to use docker-compose, created workspace and binding between host and container

Commit:
versioned /mounted jupyter config, updated README

thescientist13 commented on docker-compose.yml:
```diff
@@ -0,0 +1,21 @@
+    working_dir: /workspace
+    # since we don't have an actual process running
+    # https://stackoverflow.com/questions/38546755/docker-compose-keep-container-running
+    command: tail -F anything
```
@paulopperman / @davenamin 
not sure if we would want to make `jupyter` the default running process?

thescientist13 commented:
@paulopperman / @davenamin 
Please review, I've made some pretty big updates that I think brings this pretty to close to ready for prime time!  

Here a summary of my latest changes (full list tracked in the description)
> 4. Using `docker-compose` to help improve contributor /  CLI ergonomics
> 5. Version controlled a configuration file for the Jupyter Notebook server
> 6. Created a _workspace_ directory that can be used as a two-way mount point between the host and container
> 7. Make _workspace_ the working directory, where "starter" Jupyter Notebooks for contributors can be saved, for example

Essentially, this workspace directory can be where example Notebooks can be committed (like in #7 ) and it is also two-way mounted so whatever the user saves, will be tracked in `git` on their host machine.

Only TODO item remaining is the IP Address discrepancy between OSX / Windows, but I don't see it as a showstopper.
> Try and fix the `0.0.0.0` / `127.0.0.0` issue between OSX / Windows when loading the Jupyter Notebook server from the host

Let me know what you think, thanks for all the good feedback!

paulopperman commented:
ok, so i did some tweaking of the config file to remove security altogether (not recommended for anything in the wild.)  it works for me by going to `localhost:8889` as expected.  i also think that kicking off the server with jupyter running is a good idea, to keep the bar to entry low.  people can still `/bin/bash` if they want.

paulopperman committed:
remove token and pwd on jupyter notebook

paulopperman committed:
start jupyter notebook on container startup

davenamin commented:
hey @thescientist13 @paulopperman this looks really good! i tried it both with Owen's branch and then with Paul's changes. I agree with Paul's changes - it's definitely a lot easier to get everything going when the notebook automatically starts up and doesn't need any password.

I do have a question though: why does the docker-compose file create two containers? I was able to delete both the Dockerfile from the repository as well as the "sandbox" container in the docker-compose.yml and it worked just fine. Alternatively, if the plan is to install python packages as part of the Docker container build, maybe the "python" container in the docker-compose.yml should contain the line `build: .` instead of `image: continuumio/anaconda3:5.0.1`

paulopperman committed:
add libraries to dockerfile, point compose to NDP image

paulopperman commented:
@davenamin i just pushed a commit to #9 that uses the newportdata/datanight_python image, which is built from the dockerfile in the repo and includes the python dependencies.  dockerfile is redundant for now, and we may want to move it to its own repo at some point in the future.

also, i also just tested without sandbox (maybe that was a testing image?) and it worked fine for me.  i'm going to delete it in my branch and merge to this one.

paulopperman committed:
remove sandbox container

paulopperman committed:
Merge pull request #9 from NewportDataProject/pr-8-boot-jupyter-with-docker

Streamline docker/jupyter workflow from #8

thescientist13 commented on README.md:
```diff
@@ -43,12 +43,46 @@ We've set up a few different possible options for working with the data.  You ca
-1. Install [docker](https://www.docker.com).
-1. In the command shell, navigate to this folder run the command `docker run`
-1. Open a browser and go to `http://[CONTAINER_IP]:8888`
+**Note: For windows, use http://127.0.0.0:8889?/token=xxx**
```
@paulopperman / @davenamin 
Can this be removed now?

paulopperman replied:
yeah.  issue #13 opened for this and some other stuff.

thescientist13 commented:
@paulopperman 
> also, i also just tested without sandbox (maybe that was a testing image?) and it worked fine for me. i'm going to delete it in my branch and merge to this one.

Yeah, might have been pre-emptive, but it was mainly just trying to leverage `docker-compose` for what it does best, starting up a bunch of containers so they can talk to each other, like if wanted to add a database, like in [the example](https://hub.docker.com/u/thescientist/) I referenced in the description.

thescientist13 commented on docker-compose.yml:
```diff
@@ -0,0 +1,15 @@
+services:
+
+  python:
+    image: newportdata/datanight_python:initial
```
nice 🎉
